python/pyspark/sql/readwriter.py

```diff
@@ -303,6 +330,40 @@ def load(
         if schema is not None:
             self.schema(schema)
         self.options(**options)
+
+        # Load a Python data source
+        if isinstance(self._format, Callable):  # type: ignore[arg-type]
+            # TODO(SPARK-45560): support load() with non-empty path.
+
+            # Create an instance of the data source.
+            data_source_cls = cast(Type[DataSource], self._format)
+            data_source = data_source_cls(self._options)
+
+            # Get schema of the data source
+            schema = self._schema or data_source.schema()
```

Comment:
This is a bit hacky. For example, if we run this in Spark Connect, the schema inference would happen within Python Spark Connect client.

Reply:
Also, I don't think we can support SQL usecases if we do this here?

Reply:
Yea if a Python data source is registered and used in SQL, we need to have a separate Python process to instantiate the Python class and get the schema.

Reply:
Indeed it's a bit hacky right now. I will give this further thought when supporting `spark.read.schema(...)` in SPARK-45559. Currently, it does not work with the `schema(...)` clause.

Reply:
It will be necessary to inspect the values of the provided options for each call in order to determine which schema to generate. This is analogous to the Python UDTF `analyze` call we designed which runs on the driver.

Reply:
The `def schema()` method here can access `self.options` to generate schema based on the user inputs.